File: tests/c++/events.cpp
#include <chrono>
#include <iomanip>
#include <iostream>
#include <thread>

#include "../include/uhppoted.hpp"
#include "tests.hpp"

using namespace std;
using namespace std::this_thread;

bool getEventIndex(uhppoted &u) {
    auto index = u.get_event_index(DEVICE_ID);

    vector<result> rs = {
        result("event index", uint32_t(47), index),
    };

    return evaluate("get-event-index", rs);
}

bool setEventIndex(uhppoted &u) {
    u.set_event_index(DEVICE_ID, EVENT_INDEX);

    vector<result> rs = {};

    return evaluate("set-event-index", rs);
}

bool getEvent(uhppoted &u) {
    auto event = u.get_event(DEVICE_ID, EVENT_INDEX);

    vector<result> rs = {
        result("event index", uint32_t(51), event.index),
        result("event timestamp", string("2022-04-15 12:29:15"), event.timestamp),
        result("event type", uint8_t(6), event.eventType),
        result("event granted", true, event.granted),
        result("event door", uint8_t(3), event.door),
        result("event direction", uint8_t(1), event.direction),
        result("event card", uint32_t(8165538), event.card),
        result("event reason", uint8_t(21), event.reason),
    };

    return evaluate("get-event-index", rs);
}

bool recordSpecialEvents(uhppoted &u) {
    u.record_special_events(DEVICE_ID, true);

    vector<result> rs = {};

    return evaluate("record-special-events", rs);
}

bool listen(uhppoted &u) {
    bool listening = false;
    bool stop = false;
    vector<result> rs;

    on_event callback = +[](const struct ListenEvent e, void *userdata) {
        vector<result> *rs = (vector<result> *)userdata;

        *rs = {
            result("event controller", uint32_t(405419896), e.controller),
            result("event index", uint32_t(17), e.index),
            result("event timestamp", string("2024-07-05 12:36:45"), string(e.timestamp)),
            result("event type", uint8_t(6), e.event),
            result("event granted", true, e.granted),
            result("event door", uint8_t(2), e.door),
            result("event direction", uint8_t(1), e.direction),
            result("event card", uint32_t(10058400), e.card),
            result("event reason", uint8_t(21), e.reason),
        };
    };

    u.listen(callback, &listening, &stop, nullptr, &rs);

    sleep_for(chrono::milliseconds(100));
    for (int count = 0; count < 5 && !listening; count++) {
        cout << " ... waiting " << count << " " << (listening ? "listening" : "pending") << endl;
        sleep_for(chrono::milliseconds(500));
    }

    if (!listening) {
        throw uhppoted_exception((char *)"failed to start event listener");
    }

    sleep_for(chrono::milliseconds(1000));

    stop = true;
    for (int count = 0; count < 5 && listening; count++) {
        cout << " ... stoppping event listener " << count << " " << (listening ? "listening" : "stopped") << endl;
        sleep_for(chrono::milliseconds(100));
    }

    if (listening) {
        throw uhppoted_exception((char *)"failed to stop event listener");
    }

    return evaluate("listen", rs);
}

File: go/events.go
//go:build !debug && !tests

package main

import (
	"C"
	"fmt"
	"os"
	"os/signal"
	"syscall"

	"github.com/uhppoted/uhppote-core/uhppote"
)

func getEventIndex(uu uhppote.IUHPPOTE, index *uint32, deviceID uint32) error {
	if index == nil {
		return fmt.Errorf("invalid argument (index) - expected valid pointer")
	}

	reply, err := uu.GetEventIndex(deviceID)
	if err != nil {
		return err
	}

	*index = reply.Index

	return nil
}

func setEventIndex(uu uhppote.IUHPPOTE, deviceID uint32, index uint32) error {
	reply, err := uu.SetEventIndex(deviceID, index)
	if err != nil {
		return err
	} else if !reply.Changed {
		return fmt.Errorf("%v: failed to set event index", deviceID)
	}

	return nil
}

func getEvent(uu uhppote.IUHPPOTE, event *C.struct_Event, deviceID uint32, index uint32) error {
	if event == nil {
		return fmt.Errorf("invalid argument (event) - expected valid pointer")
	}

	e, err := uu.GetEvent(deviceID, index)
	if err != nil {
		return err
	} else if e == nil {
		return fmt.Errorf("%v: no response to get-event %v", deviceID, index)
	}

	event.timestamp = C.CString(fmt.Sprintf("%v", e.Timestamp))
	event.index = C.uint(e.Index)
	event.eventType = C.uchar(e.Type)
	event.granted = cbool(e.Granted)
	event.door = C.uchar(e.Door)
	event.direction = C.uchar(e.Direction)
	event.card = C.uint(e.CardNumber)
	event.reason = C.uchar(e.Reason)

	return nil
}

func recordSpecialEvents(uu uhppote.IUHPPOTE, deviceID uint32, enabled bool) error {
	ok, err := uu.RecordSpecialEvents(deviceID, enabled)
	if err != nil {
		return err
	} else if !ok && enabled {
		return fmt.Errorf("%v: failed to enable record special events", deviceID)
	} else if !ok && !enabled {
		return fmt.Errorf("%v: failed to disable record special events", deviceID)
	}

	return nil
}

func listen(uu uhppote.IUHPPOTE, f uhppote.Listener) error {
	os.Remove("/tmp/uhppoted-dll.pipe")

	if err := syscall.Mkfifo("/tmp/uhppoted-dll.pipe", 0666); err != nil {
		return err
	}

	if _, err := os.OpenFile("/tmp/uhppoted-dll.pipe", os.O_RDWR|os.O_CREATE|os.O_APPEND, os.ModeNamedPipe); err != nil {
		return err
	} else {
		go func() {
			q := make(chan os.Signal, 1)

			defer close(q)

			signal.Notify(q, os.Interrupt)

			uu.Listen(f, q)
		}()

		return nil
	}
}

// func listen(uu uhppote.IUHPPOTE, f uhppote.Listener) error {
// 	q := make(chan os.Signal, 1)
//
// 	defer close(q)
//
// 	signal.Notify(q, os.Interrupt)
//
// 	uu.Listen(f, q)
// 	return nil
// }
